File: ios/Plugin/AudioSessionPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(AudioSessionPlugin, "AudioSession",
           CAP_PLUGIN_METHOD(currentOutputs, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(overrideOutput, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(configure, CAPPluginReturnPromise);
)

File: ios/Plugin/AudioSessionPlugin.swift
import Foundation
import Capacitor

@objc(AudioSessionPlugin)
public class AudioSessionPlugin: CAPPlugin {
    private let implementation = AudioSession()

    override public func load() {

        implementation.load()

        implementation.interruptionObserver = { [weak self] interrupt in
            self?.notifyListeners("interruption", data: [
                "type": interrupt
            ])
        }

        implementation.routeChangeObserver = { [weak self] reason in
            self?.notifyListeners("routeChanged", data: [
                "reason": reason
            ])
        }
    }

    @objc func currentOutputs(_ call: CAPPluginCall) {
        let outputs = implementation.currentOutputs()

        call.resolve([
            "outputs": outputs
        ])
    }

    @objc func overrideOutput(_ call: CAPPluginCall) {
        let output = call.getString("type") ?? "unknown"

        implementation.overrideOutput(_output: output) { (success, message, error) -> Void in
            if error == true {
                call.reject(message ?? "")
            } else {
                call.resolve([
                    "success": success,
                    "message": message ?? ""
                ])
            }
        }
    }

    @objc func configure(_ call: CAPPluginCall) {
        let options = call.options
        implementation.configure(options: options)
        call.resolve()
    }
}
